app/src/main/java/com/hedvig/app/feature/swedishbankid/sign/SwedishBankIdSignViewModel.kt

```diff
@@ -0,0 +1,127 @@
+        object Error : ViewState()
+        object Success : ViewState()
+    }
+
+    private val _viewState = MutableStateFlow<ViewState>(
+        ViewState.StartClient
+    )
+    val viewState = _viewState.asStateFlow()
+
+    sealed class Event {
+        data class StartBankID(val autoStartToken: String) : Event()
+        object StartDirectDebit : Event()
+    }
+
+    private val _events = MutableSharedFlow<Event>(
+        extraBufferCapacity = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST,
+    )
+    val events: SharedFlow<Event> = _events
+
+    private var hasCompletedSign = false
+
+    init {
+        _events.tryEmit(Event.StartBankID(autoStartToken))
+
+        subscribeToSwedishBankIdSignStatusUseCase()
+            .onEach { response ->
+                handleNewSignStatus(response.data?.signStatus?.status?.fragments?.signStatusFragment)
+            }
+            .catch { ex ->
+                e(ex)
+                _viewState.value = ViewState.Error
+            }
+            .launchIn(viewModelScope)
+    }
+
+    fun manuallyRecheckSignStatus() {
+        viewModelScope.launch {
+            manuallyRecheckSwedishBankIdSignStatusUseCase()
+                ?.let(::handleNewSignStatus)
+        }
+    }
+
+    private fun handleNewSignStatus(signStatusFragment: SignStatusFragment?) {
+        val newViewState = toViewStateOrNull(signStatusFragment) ?: return
+        _viewState.value = newViewState
+        if (newViewState is ViewState.Success && !hasCompletedSign) {
+            hasCompletedSign = true
+            tracker.signQuotes(quoteIds)
+            loginStatusService.isViewingOffer = false
+            loginStatusService.isLoggedIn = true
+            viewModelScope.launch {
+                delay(Duration.seconds(1))
+                _events.tryEmit(Event.StartDirectDebit)
+            }
+        }
+    }
+
+    private fun toViewStateOrNull(status: SignStatusFragment?): ViewState? {
+        if (status == null) {
+            return ViewState.Error
+        }
+        return when (status.collectStatus?.status) {
+            BankIdStatus.PENDING -> {
+                when (status.collectStatus?.code) {
+                    "noClient" -> ViewState.StartClient
+                    "unknown", "userSign" -> ViewState.InProgress
+                    else -> null
+                }
+            }
+            BankIdStatus.FAILED -> {
+                when (status.collectStatus?.code) {
+                    "userCancel", "cancelled" -> ViewState.Cancelled
+                    else -> ViewState.Error
+                }
+            }
+            BankIdStatus.COMPLETE -> {
+                when (status.signState) {
+                    SignState.INITIATED, SignState.IN_PROGRESS -> null
+                    SignState.COMPLETED -> {
+                        ViewState.Success
+                    }
+                    else -> ViewState.Error
+                }
+            }
```

Comment:
Should this result in an error as well?

Reply:
The `else`-case covers `UNKNOWN__` and `null` (which is probably just a legacy schema mistake) - if we error here, then any new `BankIdStatus`es would cause error UI to show up when we perhaps do not want it to.

Reply:
Aha I see